### PR TITLE
dot func with strides impl

### PR DESCRIPTION
**Pull request by Alexander-Makaryev**
Using of cl::sycl::reduction requires compiler version 2021.4

### CODE REVIEW

**Comment by oleksandr-pavlyk** [comment]
Reduction is supported since 2021 gold, but the property initialize with identity must be new in 2021.4

Can you modify the code to provide the explicit initialization value of zero instead?

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 1]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
+    // 1D vector
+    size_t ext_input1_ndim = input1_ndim == 1 ? 2 : input1_ndim;
+    size_t* ext_input1_shape = new size_t[ext_input1_ndim];;
+    size_t* ext_input1_strides = new size_t[ext_input1_ndim];;
+    if (input1_ndim == 1)
+    {
+        ext_input1_shape[0] = 1;
+        ext_input1_shape[1] = input1_shape[0];
+        ext_input1_strides[0] = 1;
+        ext_input1_strides[1] = 1;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:224]:
```suggestion
        ext_input1_strides[1] = input1_strides[0];
```

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 2]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
+
+    // 1D vector
+    size_t ext_input1_ndim = input1_ndim == 1 ? 2 : input1_ndim;
+    size_t* ext_input1_shape = new size_t[ext_input1_ndim];;
+    size_t* ext_input1_strides = new size_t[ext_input1_ndim];;
+    if (input1_ndim == 1)
+    {
+        ext_input1_shape[0] = 1;
+        ext_input1_shape[1] = input1_shape[0];
+        ext_input1_strides[0] = 1;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:223]:
```suggestion
        ext_input1_strides[0] = 0;
```

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 3]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
+        }
+    }
+    size_t ext_input2_ndim = input2_ndim == 1 ? 2 : input2_ndim;
+    size_t* ext_input2_shape = new size_t[ext_input2_ndim];;
+    size_t* ext_input2_strides = new size_t[ext_input2_ndim];;
+    if (input2_ndim == 1)
+    {
+        ext_input2_shape[0] = input2_shape[0];
+        ext_input2_shape[1] = 1;
+        ext_input2_strides[0] = 1;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:241]:
```suggestion
        ext_input2_strides[0] = input2_strides[0];
```

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 4]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
+        ext_input2_strides[1] = 1;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:242]:
```suggestion
        ext_input2_strides[1] = 0;
```

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 5]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
+    {
+        assert(input1_size == input2_size);
+        cl::sycl::event event = dot(DPNP_QUEUE, result, input1, input2, input1_strides[0], input2_strides[0], input1_size);
         event.wait();
+        return;
+    }
+
+    // 1D vector
+    size_t ext_input1_ndim = input1_ndim == 1 ? 2 : input1_ndim;
+    size_t* ext_input1_shape = new size_t[ext_input1_ndim];;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:217]:
```suggestion
    size_t* ext_input1_shape = new size_t[ext_input1_ndim];
```

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 6]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
+    size_t* ext_input1_strides = new size_t[ext_input1_ndim];;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:218]:
```suggestion
    size_t* ext_input1_strides = new size_t[ext_input1_ndim];
```

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 7]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
-        _DataType_output* local_mem =
-            reinterpret_cast<_DataType_output*>(dpnp_memory_alloc_c(input1_size * sizeof(_DataType_output)));
+        for(size_t i = 0; i < ext_input1_ndim; ++i)
+        {
+            ext_input1_shape[i] = input1_shape[i];
+            ext_input1_strides[i] = input1_strides[i];
+        }
+    }
+    size_t ext_input2_ndim = input2_ndim == 1 ? 2 : input2_ndim;
+    size_t* ext_input2_shape = new size_t[ext_input2_ndim];;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:235]:
```suggestion
    size_t* ext_input2_shape = new size_t[ext_input2_ndim];
```

File: dpnp/backend/kernels/dpnp_krnl_common.cpp [hunk 8]
```diff
@@ -92,78 +169,195 @@ void dpnp_dot_c(void* result_out,
...
+    size_t* ext_input2_strides = new size_t[ext_input2_ndim];;
```
Comment (Alexander-Makaryev) [inline dpnp/backend/kernels/dpnp_krnl_common.cpp:236]:
```suggestion
    size_t* ext_input2_strides = new size_t[ext_input2_ndim];
```

File: dpnp/dpnp_algo/dpnp_algo_linearalgebra.pyx [hunk 9]
```diff
@@ -49,64 +49,74 @@ ctypedef void(*fptr_2in_1out_dot_t)(void * , const size_t, const size_t, const l
                                     void * , const size_t, const size_t, const long * , const long * ,
                                     void * , const size_t, const size_t, const long * , const long * )
 
+cpdef shape_type_c strides_to_vector(strides, shape):
```
Comment (Alexander-Makaryev) [inline dpnp/dpnp_algo/dpnp_algo_linearalgebra.pyx:52]:
```suggestion
cdef shape_type_c strides_to_vector(strides, shape) except *:
```

**Review by oleksandr-pavlyk — APPROVED** [review]
LGTM! Would be nice to clean up commit history.

**Pull request merged**